### PR TITLE
Allow to hide columns that have nested headers

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR fixes an issue where hiding columns causes nested headers to glitch. Several modules had to be extended with new functionalities to get the state where nested headers work correctly.
 * New modifiers are added to the _nodeModifier_ module. The nested headers tree structure can be manipulated by the new ["hideColumn"](https://github.com/handsontable/handsontable/blob/feature/issue-6879/src/plugins/nestedHeaders/stateManager/nodeModifiers/hideColumn.js#L9) and ["showColumn"](https://github.com/handsontable/handsontable/blob/feature/issue-6879/src/plugins/nestedHeaders/stateManager/nodeModifiers/showColumn.js#L9) modifiers. Within the PR, the modifiers are split into separate files for better maintainability.
 * The [_matrixGenerator_](https://github.com/handsontable/handsontable/blob/feature/issue-6879/src/plugins/nestedHeaders/stateManager/matrixGenerator.js#L35) module is rewritten to support new tree [metadata props (crossHiddenColumns)](https://github.com/handsontable/handsontable/blob/feature/issue-6879/src/plugins/nestedHeaders/stateManager/utils.js#L55). 
 * For the _IndexMapper_ module, the [_changesObservable_](https://github.com/handsontable/handsontable/tree/feature/issue-6879/src/translations/changesObservable) module is added. The module allows observing the index changes changed during HoT runtime. Currently, only the "hiding" map is supported. Thanks to that the tree does not have to build from scratch but can be modified on index change, for example:
```js
this.hot.columnIndexMapper
  .createChangesObserver('hiding')
    .subscribe((changes) => { });
```
 * To the ~Core~ IndexMapper, the new methods are added. The methods complement the IndexMapper API. So from the Handsontable instance, there is a possibility to create index maps and use them to manipulate rows/columns behavior. For instance: `hot.rowIndexMapper.createAndRegisterIndexMap` (https://github.com/handsontable/handsontable/blob/feature/issue-6879/src/translations/indexMapper.js#L225);
 * The header highlighting process is fixed. From, now on only the headers closest to the dataset are highlighted (grey). An only active highlight (blue) is applied to all header layers. <details><summary>Click here to see details</summary>This's how it works on the `develop` branch:
develop header-highlight
And how on PR's branch:
pr header-highlight
</details>  

It's worth noticing that the PR is focused on fixing the cooperation of the _NestedHeaders_ plugin and the _HidingColumns_ (and hiding map in general). The collapsible headers will be fixed in separate PR.

### Possible changes that can be treated as a braking change
The PR changes how the selection of the nested headers is held. Previously no matter what header level was clicked the selection coordinates started from `-1`.
develop
Now, the selection has been released :) It contains the row index that was actually clicked.
pr

<details>
   <summary>The same applies to the nested row headers</summary>

This's how it works on the `develop` branch:
develop rows
And how on PR's branch:
pr-rows
</details>

#### New `createAndRegisterIndexMap()` method for IndexMapper
Add ability to create and register index mapper maps. So, having the Handsontable instance it's possible to create, register, and manipulate the indexes, hiding or trimming them.

For example, I can emulate the _HiddenColumns_ plugin by creating my `hidden` map.
```js
const hiddenMap = hot.columnIndexMapper.createAndRegisterIndexMap('my-hidden-map', 'hiding');

hiddenMap.setValueAtIndex(13, true);
hiddenMap.setValueAtIndex(15, true);
hot.render();
// The Handsontable after render would hide the column at index 13 and 15
```
Or I can associate additional information like custom ID or object configuration to the specific records. While the table is altered, sorted or so I can fetch that data associated with the specific coordinates using physical indexes.
```js
const idMap = hot.rowIndexMapper.createAndRegisterIndexMap('my-id-map', 'physicalIndexToValue');

// Fill the mapper with the data
idMap.setValueAtIndex(1, { id: 'my-id-1' });
idMap.setValueAtIndex(2, { id: 'my-id-2' });
// The Handsontable while CRUD operation and so would update this mapper to be in sync with changes applied to the dataset
```
#### New `createChangesObserver()` method for IndexMapper
For the purpose of handling nested headers with hidden columns, the `ChangesObservable` is created. The method allows creating the observer that the developer can listen to. Thanks to that we can react to the index changes that happened during the code running. Currently, there is support only for the `hiding` index map type.

Usage example:
```js
this.hot.columnIndexMapper
  . createChangesObserver('hiding')
    .subscribe((changes) => {
      changes.forEach(({ op, index: columnIndex, oldValue, newValue }) => {
        if (op === 'replace' && oldValue !== newValue) {
          const actionName = newValue ? 'hide-column' : 'show-column';

          this.#stateManager.triggerColumnModification(actionName, columnIndex);
        }
      });
    });
```

#### New `unregisterAll()` method for IndexMapper
Method, when called, unregisters all index maps. The method is used in the Core, so once the table is destroyed the IndexMaps are automatically unregistered. There is no need to unregister them manually as it is done on develop branch.

#### New `beforeHighlightingColumnHeader`  and `beforeHighlightingRowHeader` hooks
The hooks allow modifying the visual indexes that are used to retrieve the header elements (TH) before they're highlighted (proper CSS class names are added). Modifying the indexes allow building a custom implementation of the nested headers feature or other features that require highlighting other DOM elements than that the rendering engine, by default, would have highlighted.

### How has this been tested?

The PR was cover with a lot of new units and E2E tests. Additionally, I tested the table manually on the latest Chrome. 

I've tested the PR using the performance lab and I did not find any symptoms that could suggest that the performance has decreased.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #6879
2. #7160 (The issue is partially fixed)
3. #6900 (the PR fixes the issue)
4. #4964 (the PR adds support for non-contiguous selection)

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### TODO
 - [x] Implement the node modifiers for hiding and showing columns (and cover the logic with unit tests);
 - [x] Refactor the _matrixGenerator_ to support new tree meta props (and cover the logic with unit tests);
 - [x] Implement the above changes into the _NestedHeaders_ plugin and cover with E2E tests;
 - [x] Check and fix (if necessary) the selection for the hidden columns;
 - [x] Finish the implementation of the Index Observability;
 - [x] Complete the documentation for new modules and API;
 - [x] Cover with tests the new API (creating index map and creating observer);
 - [x] Add support for `updateSettings`. Currently update settings reset the tree state;
 - [x]  Update TS definition file;

### CODE REVIEW

**Comment by krzysztofspilka** [comment]
Changed the title to better reflect the nature of this task as in fact it's not a bug fix but a new feature.

**Review by wszymanski — APPROVED** [review]
Good job 👏 The code looks very promising! I have just minor suggestions.

**Review by wszymanski — CHANGES_REQUESTED** [review]
Shouldn't both columns be selected when `shift` key is pressed? What about unhiding elements? Should `unhide column` be visible when `ContextMenu` plugin is enabled?

shift

**Comment by wszymanski** [comment]
It's not regression. Shouldn't unhiding work for all levels?

Screenshot 2021-04-19 at 14 31 37
Screenshot 2021-04-19 at 14 31 48

**Comment by wojciechczerniak** [comment]
> It's not regression and probably is beyond the task scope (@wojciechczerniak?). Shouldn't unhiding work for all levels?

IMO it should

File: handsontable.d.ts [hunk 1]
```diff
@@ -1996,37 +1980,55 @@ declare namespace Handsontable {
...
+      getValues(): any[],
+      getValueAtIndex(i: number): any;
+      setValues(v: any[]): void;
+      setValueAtIndex(i: number, value: any): boolean;
+      clear(): void;
+      getLength(): number;
+    }
+
+    interface ChangesEventData {
+      op: 'replace' | 'insert' | 'remove';
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable.d.ts:1992]: I think that the `op` (operation) name reflects better the essence of changes that happened in the index array. I'd leave it as it is.

File: src/core.js [hunk 2]
```diff
@@ -4021,8 +4024,6 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
       datamap.destroy();
     }
 
-    instance.rowIndexMapper = null;
-    instance.columnIndexMapper = null;
```
[the remark this answers is not in the document]
Comment (budnix) [reply src/core.js:4025]: I'm sure that if you destroy the Handsontable and remove the reference to its instance all objects would be removed. But, yeah I can agree that `null`-ing that props would decrease overall memory footprint in cases when the developer forgot to remove the reference to the destroyed instance. 

I'll revert that change 😅

File: src/plugins/nestedHeaders/__tests__/stateManager/nodeModifiers/index.unit.js [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+describe('NodeModifiers', () => {
+  describe('triggerNodeModification', () => {
+    it('should throw an error when unsupported action is used', () => {
+      expect(() => {
+        triggerNodeModification('trim');
+      }).toThrowError('The node modifier action ("trim") does not exist.');
+    });
+
+    // it('should call internal action and return modification results', () => {
```
[the remark this answers is not in the document]
Comment (budnix) [reply src/plugins/nestedHeaders/__tests__/stateManager/nodeModifiers/index.unit.js]: Good catch! 💪  This test is redundant and refers to the code before the code refactor.

Done in https://github.com/handsontable/handsontable/pull/7528/commits/2e2f8adfca651e593fbd3c0ebdeef1a32cc760c4

File: src/core.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (budnix) [reply src/core.js]: Done in https://github.com/handsontable/handsontable/pull/7528/commits/2e2f8adfca651e593fbd3c0ebdeef1a32cc760c4

**Comment by budnix** [comment]
> Shouldn't both columns be selected when `shift` key is pressed? What about unhiding elements? Should `unhide column` be visible when `ContextMenu` plugin is enabled?
> 
> shift

You're right! I'll fix that. Eagle Eye!

**Comment by budnix** [comment]
> It's not regression and probably is beyond the task scope (@wojciechczerniak?). Shouldn't unhiding work for all levels?
> 
> Screenshot 2021-04-19 at 14 31 37
> 
> Screenshot 2021-04-19 at 14 31 48

Of course, the context menu should work better when the nested headers are enabled. As you mentioned it's out of the scope for this PR. I'll create a separate issue for this.

**Comment by budnix** [comment]
> Shouldn't both columns be selected when `shift` key is pressed? What about unhiding elements? Should `unhide column` be visible when `ContextMenu` plugin is enabled?
> 
> shift

Fixed within https://github.com/handsontable/handsontable/pull/7528/commits/420da76ea13e6994f700f68c6fc9110900e091e9.

**Review by wszymanski — APPROVED** [review]

**Review by aninde — CHANGES_REQUESTED** [review]
A lot of positive changes. It was really hard for me to find something untested. With difficulty, I found one unhandled case.
But first, the good news:
- Autofill borders are displayed in full and flies everywhere where it should. https://github.com/handsontable/handsontable/issues/6900
Demo: https://jsfiddle.net/q2wkrx79/
awesomeautofill

- headers display correctly when nested headers work with hiddenColumns https://github.com/handsontable/handsontable/issues/6879
Demo: https://jsfiddle.net/2aLufc7m/
Screenshot 2021-04-23 at 17 13 44

- nestedHeaders support noncontinuous selection, even in the complex cases https://github.com/handsontable/handsontable/issues/4964
Demo: https://jsfiddle.net/mrbd3a9o/
4964-fixed

- Headers are displayed even though hiding and hiding the column https://github.com/handsontable/handsontable/issues/6879
Demo: https://jsfiddle.net/gbmtwdch/

Unfortunately, I have to mention the unhandled case.

- Bug reported in https://github.com/handsontable/handsontable/issues/7160 is fixed but only for the first use.

When you do hide row operation - horizontal scroll once that's ok. When you execute this reproduction twice, the error described in the issue comes back. 
**Reproduction:** 
1. Hide the first line (A1). 
2. Scroll horizontally to the rightmost but without clicking on the scrollbars
3. Scroll back to the beginning of the table. 
4. Hide the first line (it is now A2). Scroll right. 
The same error appears in the console as before the fix.
Demo https://jsfiddle.net/f1pscmgy/2/
7160-PR7528

**Comment by budnix** [comment]
> Unfortunately, I have to mention the unhandled case.
> 
> * Bug reported in #7160 is fixed but only for the first use.
> 
> When you do hide row operation - horizontal scroll once that's ok. When you execute this reproduction twice, the error described in the issue comes back.
> **Reproduction:**
> 
> 1. Hide the first line (A1).
> 2. Scroll horizontally to the rightmost but without clicking on the scrollbars
> 3. Scroll back to the beginning of the table.
> 4. Hide the first line (it is now A2). Scroll right.
>    The same error appears in the console as before the fix.
>    Demo https://jsfiddle.net/f1pscmgy/2/
>    7160-PR7528

@aninde Thanks for checking that. I've updated the PR description and I marked that the issue is partially fixed. Can you review the PR again without taking into consideration the mentioned issue?

**Review by aninde — APPROVED** [review]
👍

**Pull request merged**